hooks/mari/publish_mipmaps.py

```diff
@@ -91,82 +108,195 @@ def validate(self, task_settings, item):
                 self.logger.error(error_msg)
                 return False
 
-        if not super(MariPublishMipmapsPlugin, self).validate(task_settings, item):
-            return False
-
-        publish_path = sgtk.util.ShotgunPath.normalize(item.properties.get("publish_path"))
-        if not self._valid_for_mipmap_multiimage(publish_path):
-            error_msg = "Failed to find OIIO plugin or mipmap not supported for extension: %s " \
-                        "Validation failed." % os.path.splitext(publish_path)[-1]
-            self.logger.error(error_msg)
-            return False
-
         return True
 
 
-    def publish_files(self, task_settings, item, publish_path):
+    def publish_files(self, task_settings, item):
         """
         Overrides the inherited method to export out session items to the publish_path location.
 
         :param task_settings: Dictionary of Settings. The keys are strings, matching
             the keys returned in the settings property. The values are `Setting`
             instances.
         :param item: Item to process
-        :param publish_path: The output path to publish files to
         """
         publisher = self.parent
+        item.properties["mipmap_paths"] = []
 
         # get the path in a normalized state. no trailing separator, separators
         # are appropriate for current os, no double separators, etc.
-        target_path = sgtk.util.ShotgunPath.normalize(publish_path)
-
-        try:
-            # ensure the publish folder exists:
-            publish_folder = os.path.dirname(target_path)
-            ensure_folder_exists(publish_folder)
-
-            # get all associated files if this is a file sequence
-            source_path = item.parent.properties.publish_path
-            seq_path = publisher.util.get_frame_sequence_path(source_path)
-            if seq_path:
-                src_files = publisher.util.get_sequence_path_files(source_path)
-            else:
-                src_files = [source_path]
+        for extension in item.properties["mipmap_extensions"]:
+            target_path = sgtk.util.ShotgunPath.normalize(
+                self._get_publish_path_with_ext(item, task_settings, extension))
 
-            # write mipmaps for each file in the sequence
-            mipmap_paths = self.create_mipmaps_for_seq(src_files, target_path)
+            try:
+                # ensure the publish folder exists:
+                publish_folder = os.path.dirname(target_path)
+                ensure_folder_exists(publish_folder)
 
-        except Exception as e:
-            raise TankError("Failed to publish file for item '%s': %s" % (item.name, str(e)))
+                # Find source file from current item.properties.sg_publish_data_list
+                # since product of publish_texture is the source file, and the item
+                # is shared between the plugins
+                source_path = item.properties["sg_publish_data_list"][0]["path"]["local_path"]
```

Comment:
This feels odd, but if item is to be shared between publish_texture and publish_mipmap, we can't rely on parent item, and have to use/modify the same item's properties. I don't like this.

Reply:
A good reason to store the result of `self.publish_files()` :) We should maybe also segregate by task...i.e. in publish.py:

```python
        self.publish_files(task_settings, item, publish_path)
```
becomes:
```python
        publish_files = self.publish_files(task_settings, item, publish_path)
        if "publish_files" not in item.properties:
            item.properties.publish_files = {}
        item.properties.publish_files{task.name} = self.publish_files(task_settings, item, publish_path)
```

Reply:
This would also remove the need for `item.properties["mipmap_paths"]` since you could reference `item.properties.publish_paths["Publish Mipmaps"]`

Reply:
Okay, so I tried making this and `sg_publish_data_list` task specific properties, but that doesn't seem to be enough.
I'm running into errors with item.properties.fields["extension"] being carried over between plugins.
I can find a way around this particular error, but in general, "item x task" should have its own set of properties...? Dealing with it on a per property basis seems like setting ourselves up for future failures. Feels like a major refactor (I would like "item x task" to be a separate object, but that may make merging with shotgun a different headache), but let me know what you think.

Reply:
I now realized that task_specific properties are already supported via `item.local_properties`. Use that instead of `item.properties`. We should probably do the same for "extension". They then also provide the convenience method `get_property()` to first check `local_properties` and if not found `properties`.